tools/java/runfiles/AutoBazelRepositoryProcessor.java

```diff
@@ -0,0 +1,104 @@
+import javax.lang.model.element.TypeElement;
+import javax.tools.Diagnostic.Kind;
+
+@SupportedAnnotationTypes("com.google.devtools.build.runfiles.AutoBazelRepository")
+@SupportedOptions("bazel.repository")
+public final class AutoBazelRepositoryProcessor extends AbstractProcessor {
+
+  @Override
+  public SourceVersion getSupportedSourceVersion() {
+    return SourceVersion.latestSupported();
+  }
+
+  @Override
+  public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+    annotations.stream()
+        .flatMap(element -> roundEnv.getElementsAnnotatedWith(element).stream())
+        .map(element -> (TypeElement) element)
+        .forEach(this::emitClass);
+    return true;
+  }
+
+  private void emitClass(TypeElement annotatedClass) {
+    // This option is always provided by JavaBuilder.
+    String repositoryName = processingEnv.getOptions()
+        .getOrDefault("bazel.repository", "for testing only");
+    if (repositoryName == null) {
+      // javac translates '-Abazel.repository=' into a null value.
+      // https://github.com/openjdk/jdk/blob/7a49c9baa1d4ad7df90e7ca626ec48ba76881822/src/jdk.compiler/share/classes/com/sun/tools/javac/processing/JavacProcessingEnvironment.java#L651
+      repositoryName = "";
+    }
+
+    // For a nested class Outer.Middle.Inner, generate a class with simple name
+    // AutoBazelRepository_Outer_Middle_Inner.
+    // Note: There can be collisions when local classes are involved, but since the definition of a
+    // class depends only on the containing Bazel target, this does not result in ambiguity.
+    List<String> nestedClassNames = Stream.iterate(annotatedClass,
+            element -> element instanceof TypeElement,
+            Element::getEnclosingElement)
+        .map(Element::getSimpleName)
+        .map(Name::toString)
+        .collect(toList());
+    Collections.reverse(nestedClassNames);
+    String generatedClassSimpleName = Stream.concat(Stream.of("AutoBazelRepository"),
+        nestedClassNames.stream()).collect(joining("_"));
+
+    String generatedClassPackage = processingEnv.getElementUtils().getPackageOf(annotatedClass)
+        .getQualifiedName().toString();
+
+    String generatedClassName = generatedClassPackage.isEmpty() ? generatedClassSimpleName
+        : generatedClassPackage + "." + generatedClassSimpleName;
+
+    try (PrintWriter out = new PrintWriter(
+        processingEnv.getFiler().createSourceFile(generatedClassName).openWriter())) {
+      out.printf("package %s;\n", generatedClassPackage);
+      out.printf("\n");
+      out.printf("class %s {\n", generatedClassSimpleName);
+      out.printf("  /**\n");
+      out.printf("   * The canonical name of the repository containing the Bazel target that\n");
+      out.printf("   * compiled {@link %s}.\n", annotatedClass.getQualifiedName().toString());
+      out.printf("   */\n");
+      out.printf("  static final String BAZEL_REPOSITORY = \"%s\";\n", repositoryName);
```

Comment:
@cushon Given that the class is called `AutoBazelRepository_X`, should I maybe replace this with `NAME`? More concise and less repetitive.